### PR TITLE
SerializeLoc: serialize basic decl source location information to .swiftsourceinfo file

### PR DESCRIPTION
**Pull request by nkcsgexi**
After setting up the .swiftsourceinfo file, this patch starts to actually serialize and de-serialize source locations for declaration. The binary format of .swiftsourceinfo currently contains these three records:

BasicDeclLocs: a hash table mapping from a USR ID to a list of basic source locations. The USR id could be retrieved from the following DeclUSRs record using an actual decl USR. The basic source locations include a file ID and the results from Decl::getLoc(), ValueDecl::getNameLoc(), Decl::getStartLoc() and Decl::getEndLoc(). The file ID could be used to retrieve the actual file name from the following SourceFilePaths record. Each location is encoded as a line:column pair.

DeclUSRS: a hash table mapping from USR to a USR ID used by location records.

SourceFilePaths: a hash table mapping from a file ID to actual file name.

BasicDeclLocs should be sufficient for most diagnostic cases. If additional source locations are needed, we could always add new source location records without breaking the backward compatibility. When de-serializing the source location from a module-imported decl, we calculate its USR, retrieve the USR ID from the DeclUSRS record, and use the USR ID to look up the basic location list in the BasicDeclLocs record.

For more details about .swiftsourceinfo file: https://forums.swift.org/t/proposal-emitting-source-information-file-during-compilation

rdar://56167685

### CODE REVIEW

File: test/Serialization/comments.swift
Comment (nkcsgexi) [inline test/Serialization/comments.swift]: @rintaro as you have suggested, a test case for `#sourceLocation` is added here.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please test compiler performance

**Comment by nkcsgexi** [comment]
@swift-ci Please Build Toolchain macOS Platform

File: include/swift/AST/RawComment.h
Comment (jrose-apple) [inline include/swift/AST/RawComment.h]: I don't think we have any decls where Loc and NameLoc are different. Is this just to account for decls without names?

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Hmm, I guess this is future-proofing for other tables keyed by USR, so that the string doesn't get repeated in the keys of every table? It does seems like if you have an integer ID you don't need a hash table any more, though.

File: include/swift/AST/RawComment.h
Comment (nkcsgexi) [reply include/swift/AST/RawComment.h]: We have separate APIs `getLoc()` and `getNameLoc()` for `ValueDecl`, so they could potentially be different. IMO, we shouldn't hardcode that they are usually identical to the format.

File: lib/Serialization/ModuleFormat.h
Comment (jrose-apple) [inline lib/Serialization/ModuleFormat.h]: I thought you decided these *weren't* going to be considered a stable format, since they aren't installed? As such I'm not sure you need to put them after the moduledoc numbers.

Comment (jrose-apple) [reply lib/Serialization/ModuleFormat.h]: If they are a stable format I'm a little more concerned about the format of these hash tables.

File: lib/Serialization/SerializedModuleLoader.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializedModuleLoader.cpp]: This seems like duplicated code; can you factor out the common parts instead?

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Missing the `-*- c++ -*-`.

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Is it the DECL_LOCS block or the SOURCEINFO block?

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Why the jump to 96?

File: lib/Serialization/SerializeDoc.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: Typo: Coloumn

File: lib/Serialization/SerializeDoc.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: If the length is constant you actually don't have to serialize it at all; returning it is sufficient.

File: lib/Serialization/SerializeDoc.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: Inconsistent about the semicolons here. (I liked the lambda better, honestly.)

**Review by jrose-apple — COMMENTED** [review]
A storm of comments on just the first commit, sorry.

File: include/swift/AST/RawComment.h
Comment (jrose-apple) [reply include/swift/AST/RawComment.h]: I'll flip it around: there are zero cases where `getLocFromSource()` does not return `getNameLoc()` if there is a `getNameLoc()`, and we could put that in the AST verifier.

**Review by davidungar — APPROVED** [review]
I just looked over it quickly, but nothing jumped out at me to comment upon. (I'll miss @jrose-apple 's reviews!) Let me know if you want another pair of eyes to go over it in depth.

**Comment by nkcsgexi** [comment]
@jrose-apple The most recent commit refactored the data structure as we discussed to use only one hash-table. Could you take another look?

File: lib/Serialization/SerializedModuleLoader.cpp
Comment (nkcsgexi) [reply lib/Serialization/SerializedModuleLoader.cpp]: hmm, it seems we have similar duplications for other two module inputs too. Do you have any ideas where we should move it to?

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/Serialization/SerializedModuleLoader.h
Comment (brentdax) [inline include/swift/Serialization/SerializedModuleLoader.h]: These now-three files are passed together so often that I'm beginning to think we should have an abstraction to represent this grouping. `ModuleFiles<StringRef> ModuleFilenames`, `ModuleFiles<std::unique_ptr<llvm::MemoryBuffer> *> ModuleBuffers`.

File: include/swift/AST/RawComment.h
Comment (brentdax) [inline include/swift/AST/RawComment.h]: Would it make sense to assert somewhere that we never create a `LineColumn` with a zero `Line` but a nonzero `Column`, or vice versa? I worry about one of these fields ending up with garbage in it without us noticing because the other is zeroed, and then something reading that garbage and doing the wrong thing.

**Review by beccadax — APPROVED** [review]
I'm not going to pretend I had the attention span to read every line in detail, but it looked alright in my skim. If there's something specific you want me to examine more minutely, please let me know.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/AST/RawComment.h
Comment (nkcsgexi) [reply include/swift/AST/RawComment.h]: hmm, not sure it worths it. We plan to use `LineColumn` here as an internal implementation detail just like `CommentInfo` defined above so users won't directly interact with this type. I'll open future PRs to hook `LineColumn` up with AST APIs like `getLoc()`, etc.

File: lib/Frontend/Frontend.cpp
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: Ooh, nice cleanup! Even better than what I had. :-)

File: lib/Frontend/Frontend.cpp
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: Nitpicks: this should be a SmallString too, and it should have a lowerCamelCase name, and I don't love that the name is described in terms of the *other* path.

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Nitpick: indentation

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Nitpick: 80 cols

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: \*sigh\* Technically this kind of reinterpret_cast is illegal in C++, please use the `endian::readNext` API anyway. (Also, it's important if we ever cross platform boundaries with endianness! For some reason!)

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Nitpick: You can omit this last argument since DeclUSRTableInfo has a default constructor.

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: No error handling?

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Same deal, use explicit `endian::readNext` operations, please.

File: lib/Serialization/ModuleFile.h
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.h]: This comment and the one below are no longer accurate.

Comment (jrose-apple) [reply lib/Serialization/ModuleFile.h]: Also, please keep all the stored properties together at the top of the class.

File: lib/Serialization/ModuleFormat.h
Comment (jrose-apple) [reply lib/Serialization/ModuleFormat.h]: We talked about this in person. The plan is to attempt stability, but the consequences aren't so bad if we fail.

File: lib/Serialization/SerializedModuleLoader.cpp
Comment (jrose-apple) [reply lib/Serialization/SerializedModuleLoader.cpp]: I mean, here is fine, just generalize it.

File: lib/Frontend/Frontend.cpp
Comment (jrose-apple) [inline lib/Frontend/Frontend.cpp]: I don't think we need to do the "Project" thing for the partial modules.

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Where does this number come from? :-)

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Don't forget to update the comment!

File: lib/Serialization/SourceInfoFormat.h
Comment (jrose-apple) [inline lib/Serialization/SourceInfoFormat.h]: Not very extensible, but probably good enough.

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: If you want to be extra future-proof, check if `BasicDeclLocsData` is empty first. That would indicate a future version of the swiftsourceinfo file format that doesn't provide BasicDeclLocsData anymore. (Ditto for the other lookups.)

**Review by jrose-apple — COMMENTED** [review]
More comments on everything except SerializeDoc.cpp.

File: lib/Frontend/Frontend.cpp
Comment (jrose-apple) [reply lib/Frontend/Frontend.cpp]: Actually, I don't think we need to do the "Project" thing for the partial modules.

File: lib/Serialization/SourceInfoFormat.h
Comment (nkcsgexi) [reply lib/Serialization/SourceInfoFormat.h]: I made these number up. Is there a recommended way?

Comment (harlanhaskins) [reply lib/Serialization/SourceInfoFormat.h]: Concept: use the UTF8 bytes of 👩‍⚕️

File: lib/Serialization/ModuleFile.cpp
Comment (nkcsgexi) [reply lib/Serialization/ModuleFile.cpp]: Failing to load `.swiftsourceinfo` isn't a very big deal and probably is the default situation when loading a shipped Swift module. So I don't think it worths error handling.

File: lib/Serialization/SourceInfoFormat.h
Comment (harlanhaskins) [reply lib/Serialization/SourceInfoFormat.h]: Concept: use the UTF8 bytes of 📒

File: lib/Serialization/SerializeDoc.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: These two sentences are a little contradictory as written.

**Review by jrose-apple — COMMENTED** [review]
Comments on the serialization side, mostly minor.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Serialization/SerializeDoc.cpp
Comment (nkcsgexi) [reply lib/Serialization/SerializeDoc.cpp]: Moving this logic to `SourceFile:: getBasicLocsForDecl` raises a question of how to manage the memory of the source file paths in the ASTContext, which probably doesn't worth it.

Comment (jrose-apple) [reply lib/Serialization/SerializeDoc.cpp]: `SourceManager::getDisplayNameForLoc` is already returning a StringRef, so we should be okay.

Comment (nkcsgexi) [reply lib/Serialization/SerializeDoc.cpp]: We need to call `llvm::sys::fs::make_absolute` on the path, so the StringRef isn't sufficient.

Comment (jrose-apple) [reply lib/Serialization/SerializeDoc.cpp]: Hmm. Can we do that at the point of serialization instead of when returning a value? It's still not great that we do it at all.

Comment (nkcsgexi) [reply lib/Serialization/SerializeDoc.cpp]: Ok, the most recent commit moved this logic to `SourceFile::getBasicLocsForDecl()`.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: lib/Serialization/SourceInfoFormat.h
Comment (nkcsgexi) [reply lib/Serialization/SourceInfoFormat.h]: I end up using 🏎️as the magic number :)

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test compiler performance

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
Talked with @jrose-apple about this, we plan to merge it for now to unblock progress. Future code review comments, if any, will be addressed by separate PRs.

**Comment by compnerd** [comment]
I think that this may have caused a regression on Windows: https://github.com/apple/swift-corelibs-foundation/pull/2521/files

File: lib/Serialization/SerializeDoc.cpp
Comment (jrose-apple) [inline lib/Serialization/SerializeDoc.cpp]: Performance nitpick: this will do 2-3 hash lookups of the text, and while that's not *super* important, you can do better with something like this:

```cxx
auto iterAndIsNew = IndexMap.insert({Text, Buffer.size()});
if (iterAndIsNew.second) {
  Buffer.append(Text);
  Buffer.push_back('\0');
}
return (*IterAndIsNew.first).getValue();
```

File: lib/Serialization/ModuleFile.cpp
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Funky indentation here.

**Review by jrose-apple — COMMENTED** [review]
Further comments, none of which are too serious.

**Pull request merged**
